rasp/tor_engine.py

```diff
@@ -1,57 +1,109 @@
+            can also be set with the ``RASP_TOR_CONTROL_PASSWORD``
+            environment variable.
+        control_password (str, optional): Password to protect control
+            port usage,
+            can also be set with the ``RASP_TOR_CONTROL_PORT``
+            environment variable.
+     """
+    def __init__(self,
+                 headers=None,
+                 address=None,
+                 port=None,
+                 control_port=None,
+                 control_password=None):
+        super(TorEngine, self).__init__(headers)
 
-        proxy_opener = urllib.request.build_opener(self.proxy_handler)
-        urllib.request.install_opener(proxy_opener)
+        self.address = (
+            address
+            or os.environ.get('RASP_TOR_ADDRESS')
+            or '127.0.0.1'
+        )
+        self.port = (
+            port
+            or os.environ.get('RASP_TOR_PORT')
+            or 9050
+        )
+        self.control_password = (
+            control_password
+            or os.environ.get('RASP_TOR_CONTROL_PASSWORD')
+            or getpass.getpass("Tor control password: ")
+        )
+        self.control_port = (
+            control_port
+            or os.environ.get('RASP_TOR_CONTROL_PORT')
+            or 9051
+        )
+
+        proxy_uri = 'socks5://{}:{}'.format(self.address, self.port)
+        proxies = {'http': proxy_uri, 'https': proxy_uri}
+        self.session.proxies.update(proxies)
 
     def __copy__(self):
         return TorEngine(
-            self.pw,
-            self.control,
-            self.signal,
-            self.proxy_handler,
             self.data,
-            self.headers
+            self.headers,
+            self.address,
+            self.port,
+            self.control_port,
+            self.control_password,
         )
 
-    def send_signal(self):
-        conn = stem.connection.connect(
-            control_port=self.control,
-            password=self.pw
-        )
-        conn.signal(self.signal)
-        conn.close()
-
-    def get_page_source(self, url):
-        try:
-            self.send_signal()
-            req = urllib.request.Request(url, self.data, self.headers)
-            res = urllib.request.urlopen(req)
-
-            try:
-                return Webpage(url, str(res.read()))
-            except http.client.IncompleteRead:
-                return
-        except urllib.error.HTTPError as e:
-            time.sleep(2)
+    def clean_circuits(self):
+        """Refreshes the mapping of nodes we connect through.
+
+        When a new path from the user to the destination is required, this
+        method refreshes the node path we use to connect. This gives us a new IP
+        address with which the destination sees."""
+        kwargs = {'address': self.address, 'port': self.control_port}
+        with Controller.from_port(**kwargs) as controller:
+            controller.authenticate(password=self.control_password)
```

Comment:
The problem with relying on `__del__` is that there's no promise when it will be called, and is hidden from the control flow of the program. The `__enter__` and `__exit__` of a context manager is explicit, and will ensure leaks won't happen.

@danlrobertson You're right that with CPython, the GC is keeping track of the refcount and when it hits 0 the object is destructed, but we can't guarantee that the user will be using CPython.  The refcount is also based on scope, and sometimes things stay in scope for longer than can be determined by the static control flow of the code (think attributes inside of a decorator). 

The biggest concern for me is that when an exception is raised, `__del__` straight up won't run.  This is the second major reason why we have context managers to begin with.  I think we should go with @wdm0006 suggestion and utilize both context managers, and the ability to close out the connection manually (which someone could do in a try/except clause).

Also, keep in mind that if we find ourselves with an error raised in `__init__`, the object won't complete initialization, thus won't ever have a non-zero refcount and will never have its `__del__` method called.

Basically, we shouldn't use `__del__`.